gammapy/modeling/fit.py

```diff
@@ -113,6 +115,17 @@ def run(self, optimize_opts=None, covariance_opts=None):
         # back or how to form the FitResult object.
         optimize_result._success = optimize_result.success and covariance_result.success
 
+        # set sub-covariance matrix for each model
+        for dataset in self.datasets.datasets:
+            if isinstance(dataset.model, SkyModels):
+                for model in dataset.model.skymodels:
+                    pars = model.parameters
+                    pars.set_subcovar_from_parameters(self._parameters)
+                    model.spatial_model.parameters.set_subcovar_from_parameters(pars)
```

Comment:
I think this is incomplete, because the model can be a `SkyDiffuseCube` as well. Also setting the covariance for the `background_model` is missing. Instead I would propose a different approach:

- Add a `Datasets._model_list` read only property, which walks all datasets and collects all model components from the individual datasets in a list.

- Then the code above reduces to:
```
for model in datasets._model_list:
    model.parameters.covariance = self._parameters.get_subcovariance(model.parameters)
```

Reply:
This would also be a first step towards a global models objects we've discussed in the past week...